feat(aws): add check `secretsmanager_has_restrictive_resource_policy`

kagahd opened the pull request:
### Context

This feature request offers a new AWS check `secretsmanager_has_restrictive_resource_policy`. 
The check evaluates resource-based policies for AWS Secrets Manager secrets.

### Description

To pass the check, resource-based policies attached to secrets must meet **all** of the following criteria:

## 1. Explicit "Deny" Statement for Unauthorized Principals

The resource policy must include an explicit `Deny` statement that applies to **all principals**, **all actions**, and **all resources**, unless exceptions are defined in the `StringNotEquals` Condition block.

**Example**

```
{
  "Sid": "DenyUnauthorizedPrincipals",
  "Effect": "Deny",
  "Principal": "*",
  "Action": "*",
  "Resource": "*",
  "Condition": {
    "StringNotEquals": {
      "aws:PrincipalArn": [
        "<ASSUMED-ROLE-ARN-FOR-SECURY-AUDIT>",
        "arn:aws:iam::<YOUR_AWS_ACCOUNT_ID>:role/SomeOtherRole"
      ]
    }
  }
}
...<further required policy statements>...
```
This structure ensures that unauthorized access is denied across the board.

**Principal: `"*"`**  
Applies the denial to all entities, internal or external, preventing unauthorized principals from accessing the secret.

**Action: `"*"` or `"secretsmanager:*"`**  
Covers all possible actions, ensuring no unauthorized operation—directly or indirectly—can be performed on the secret. Using `"secretsmanager:*"` is acceptable to limit the denial to actions related specifically to Secrets Manager.

**Resource: `"*"`**  
Because resource-based policies are inherently tied to the secret they protect, this field can either use a wildcard (`"*"` ) for broader applicability or the exact ARN of the secret for stricter targeting.

The `StringNotEquals` Condition block refines the denial by allowing **exceptions**, such as specific roles defined in `aws:PrincipalArn`.

One **mandatory** exception must be given to `<ASSUMED-ROLE-ARN-FOR-SECURY-AUDIT>` because this role, which is deployed in every organization's AWS account, must be allowed to audit the secret by the IT security team.

All Principal ARNs that will have their own policy entry must also be listed here. Otherwise, this rule would already block any access attempts.

### Using wildcards in Principal ARNs

The check allows using an `ArnNotLike` condition with key `aws:PrincipalArn` within the explicit `Deny` statement for unauthorized principals. The role name must contain **at least 12 characters before the wildcard**.

**Example**

```
{
  "Sid": "DenyUnauthorizedPrincipals",
  "Effect": "Deny",
  "Principal": "*",
  "Action": "*",
  "Resource": "*",
  "Condition": {
    "StringNotEquals": {
      "aws:PrincipalArn": [
        "<ASSUMED-ROLE-ARN-FOR-SECURY-AUDIT>",
        "arn:aws:iam::<YOUR_AWS_ACCOUNT_ID>:role/SomeOtherRole"
      ]
    },
    "ArnNotLike": {
      "aws:PrincipalArn": [
        "arn:aws:iam::<YOUR_AWS_ACCOUNT_ID>:role/<PREFIX>*",
        "arn:aws:iam::<YOUR_AWS_ACCOUNT_ID>:role/<PREFIX>*"
      ]
    }
  }
}
...<further required policy statements>...
```


### Excluding AWS services from the Deny

If AWS services must also be excluded from the Deny statement, the `StringNotEqualsIfExists` condition operator must be used instead of `StringNotEquals`. This ensures that both principals and services can be excluded even though only one of them may exist in the request.

Additionally, a `Null` condition block must be included which contains both `aws:PrincipalArn` and `aws:PrincipalServiceName` with value `"true"`. This ensures that requests missing both keys are denied.

**Example**

```
{
  "Sid": "DenyUnauthorizedPrincipals",
  "Effect": "Deny",
  "Principal": "*",
  "Action": "*",
  "Resource": "*",
  "Condition": {
    "StringNotEqualsIfExists": {
      "aws:PrincipalArn": [
        "<ASSUMED-ROLE-ARN-FOR-SECURY-AUDIT>",
        "arn:aws:iam::<YOUR_AWS_ACCOUNT_ID>:role/SomeOtherRole"
      ],
      "aws:PrincipalServiceName": "appflow.amazonaws.com"
    },
    "Null": {
      "aws:PrincipalArn": "true",
      "aws:PrincipalServiceName": "true"
    }
  }
}
...<further required policy statements>...
```

## 2. Restrict Access Outside the Organization

The second **mandatory** `Deny` statement ensures that principals **outside the AWS Organization** are denied access to the secret.
```
{
  "Sid": "DenyOutsideOrganization",
  "Effect": "Deny",
  "Principal": "*",
  "Action": "*",
  "Resource": "*",
  "Condition": {
    "StringNotEquals": {
      "aws:PrincipalOrgID": "<YOUR_ORG_ID>"
    }
  }
}
...<further required policy statements>...
```

The `StringNotEquals` condition ensures that the denial applies to any request coming from outside of the organization, as identified by the `aws:PrincipalOrgID` condition key.

The value(s) of `aws:PrincipalOrgID` can be configured via the key `organizations_trusted_ids` in prowler's [config.yaml](https://github.com/prowler-cloud/prowler/blob/master/prowler/config/config.yaml).

If the AWS account is **not part of an organization**, this value can be left empty in the configuration. In that case, the check `secretsmanager_has_restrictive_resource_policy` will **not require** the `DenyOutsideOrganization` statement.

### Dealing with AWS Services

AWS services (e.g. `"appflow.amazonaws.com"`) do not send a `PrincipalOrgID` in their requests. Therefore the service name must also be added to the condition block.

**Example**

```
{
  "Sid": "DenyOutsideOrganization",
  "Effect": "Deny",
  "Principal": "*",
  "Action": "*",
  "Resource": "*",
  "Condition": {
    "StringNotEquals": {
      "aws:PrincipalOrgID": "<YOUR_ORG_ID>",
      "aws:PrincipalServiceName": "appflow.amazonaws.com"
    }
  }
}
...<further required policy statements>...
```

This statement alone does **not** guarantee that the service runs within the same AWS account or organization.  
Therefore the service must additionally be restricted in its `Allow` statement using `aws:SourceAccount` (see section **4. Allow Statement for AWS services**).

## 3. Deny Statement to Permit Only Specific Actions for Designated Principals

These `Deny` statements ensure that specific principals are restricted to performing **only specific actions**.

These statements are **optional**, but if omitted the principals listed in `DenyUnauthorizedPrincipals` will effectively be denied all actions.

Any principal listed in the `StringNotEquals` condition of `DenyUnauthorizedPrincipals` must have a corresponding `Deny` statement with a tailored `NotAction` clause.

Wildcards (`*`) are **not allowed** in `NotAction`. Refer to the supported [AWS Secrets Manager actions](https://docs.aws.amazon.com/secretsmanager/latest/apireference/API_Operations.html).

The IT security team must be able to audit the resource-based policy. Therefore the principal `<ASSUMED-ROLE-ARN-FOR-SECURY-AUDIT>` must be permitted the actions:

- `secretsmanager:DescribeSecret`
- `secretsmanager:GetResourcePolicy`

**Example**

```
{
  "Sid": "AllowAuditPolicyRead",
  "Effect": "Deny",
  "Principal": {
    "AWS": "<ASSUMED-ROLE-ARN-FOR-SECURY-AUDIT>"
  },
  "NotAction": [
    "secretsmanager:DescribeSecret",
    "secretsmanager:GetResourcePolicy"
  ],
  "Resource": "*"
},
{
  "Sid": "AllowSecretAccessForSomeOtherRole",
  "Effect": "Deny",
  "Principal": {
    "AWS": "arn:aws:iam::<YOUR_AWS_ACCOUNT_ID>:role/SomeOtherRole"
  },
  "NotAction": [
    "secretsmanager:GetSecretValue"
  ],
  "Resource": "*"
}
...<further required policy statements>...
```

### Using wildcards for principals

The check allows using `ArnLike` with `aws:PrincipalArn` to define the allowed actions for principals that were specified via `ArnNotLike` in the `DenyUnauthorizedPrincipals` statement.

The role name must again contain **at least 12 characters before the wildcard**.

**Example**

```
{
  "Sid": "AllowReadOnlyRolePrefixAccess",
  "Effect": "Deny",
  "Principal": "*",
  "NotAction": [
    "secretsmanager:GetSecretValue",
    "secretsmanager:DescribeSecret"
  ],
  "Resource": "*",
  "Condition": {
    "ArnLike": {
      "aws:PrincipalArn": [
        "arn:aws:iam::<YOUR_AWS_ACCOUNT_ID>:role/<PREFIX>*",
        "arn:aws:iam::<YOUR_AWS_ACCOUNT_ID>:role/<PREFIX>*"
      ]
    }
  }
}
...<further required policy statements>...
```

## 4. Allow Statement for AWS Services

AWS services require an explicit `Allow` statement to perform actions on the secret.

Two requirements apply:

1. **No wildcard (`*`) in any Action**
2. The AWS **SourceAccount must match the current account** to prevent access from services running in external AWS accounts. Additional restrictive conditions (e.g. `ArnLike` on `aws:SourceArn`) are accepted alongside `aws:SourceAccount` and will not cause the check to fail.

Both simple service principals (e.g. `appflow.amazonaws.com`) and regionalized service principals (e.g. `logs.eu-central-1.amazonaws.com`) are supported.

**Example**

```
{
  "Sid": "AllowAppFlowAccess",
  "Effect": "Allow",
  "Principal": {
    "Service": "appflow.amazonaws.com"
  },
  "Action": [
    "secretsmanager:GetSecretValue",
    "secretsmanager:PutSecretValue",
    "secretsmanager:DeleteSecret",
    "secretsmanager:DescribeSecret",
    "secretsmanager:UpdateSecret"
  ],
  "Resource": "*",
  "Condition": {
    "StringEquals": {
      "aws:SourceAccount": "<YOUR_AWS_ACCOUNT_ID>"
    }
  }
}
...<further required policy statements>...
```

### Cross-account access

Cross-account access to secrets is **not allowed**.  
Resource-based policies that allow cross-account access will cause the check to **fail on purpose**. This behavior is intentional to ensure that any cross-account access is made explicitly visible during reviews and compliance checks, since from an IT Security perspective it expands the trust boundary and therefore requires deliberate assessment and clear justification.

## Complete Resource-Based Policy Example

Below is a full example resource policy for scenarios where **only IAM principals require access**.

```
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "DenyUnauthorizedPrincipals",
      "Effect": "Deny",
      "Principal": "*",
      "Action": "*",
      "Resource": "*",
      "Condition": {
        "StringNotEquals": {
          "aws:PrincipalArn": [
            "<ASSUMED-ROLE-ARN-FOR-SECURY-AUDIT>",
            "arn:aws:iam::<YOUR_AWS_ACCOUNT_ID>:role/SomeOtherRole"
          ]
        }
      }
    },
    {
      "Sid": "DenyOutsideOrganization",
      "Effect": "Deny",
      "Principal": "*",
      "Action": "*",
      "Resource": "*",
      "Condition": {
        "StringNotEquals": {
          "aws:PrincipalOrgID": "<YOUR_ORG_ID>"
        }
      }
    },
    {
      "Sid": "AllowAuditPolicyRead",
      "Effect": "Deny",
      "Principal": {
        "AWS": "<ASSUMED-ROLE-ARN-FOR-SECURY-AUDIT>"
      },
      "NotAction": [
        "secretsmanager:DescribeSecret",
        "secretsmanager:GetResourcePolicy"
      ],
      "Resource": "*"
    },
    {
      "Sid": "AllowSecretAccessForSomeOtherRole",
      "Effect": "Deny",
      "Principal": {
        "AWS": "arn:aws:iam::<YOUR_AWS_ACCOUNT_ID>:role/SomeOtherRole"
      },
      "NotAction": [
        "secretsmanager:GetSecretValue"
      ],
      "Resource": "*"
    }
  ]
}
```

---

### Unit Tests

As shown in `secretsmanager_has_restrictive_resource_policy_test.py`, several unit tests were created.  
Three of them alone cover already **105 different scenarios** by modifying parts of an otherwise valid resource policy to ensure the check correctly **fails**.
- `test_secretsmanager_policies_for_principals` = 73 test cases
- `test_secretsmanager_policies_for_principals_and_services` = 29 test cases
- `test_policy_with_arn_not_like` = 3 test cases

Each negative test case is followed by a **counter-test** that removes the invalid change to ensure the policy **passes**, verifying that the test logic itself is correct.

`@pytest.mark.parametrize` is used to reduce redundancy while maintaining readability.

Additionally, **3 regression tests** cover order-independent cross-account detection, service Allow with extra restrictive conditions, and regionalized service principals.

---

### Checklist

- Are there new checks included in this PR? **Yes**
    - If so, do we need to update permissions for the provider? **No**
- [x] Review if the code is being covered by tests.
- [x] Review if code is documented following the Google Python style guide.
- [x] Review if backport is needed.
- [x] Review if changes to `README.md` are required.

#### API

- [x] Verify if API specs need to be regenerated.
- [x] Check if version updates are required.
- [x] Ensure new entries are added to `CHANGELOG.md` if applicable.

---

### License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

kagahd committed:
feat(aws) add check secretsmanager_has_restrictive_resource_policy

kagahd committed:
feat(aws) add check secretsmanager_has_restrictive_resource_policy

jfagoagas commented on tests/providers/aws/services/secretsmanager/secretsmanager_has_restrictive_resource_policy/secretsmanager_has_restrictive_resource_policy_test.py:
I think https://docs.pytest.org/en/stable/example/parametrize.html would be ideal for this tests, let me know what you think 🙌

kagahd replied:
Oh, I've never heard of it but seems that's really nice! Let me adapt the code asap.

kagahd committed:
feat(aws) use pytest.mark.parametrize

kagahd committed:
feat(aws) reset_moto

kagahd committed:
make black & flake happy

kagahd committed:
implicit reset_moto

kagahd committed:
scope function instead of class

kagahd committed:
consider services which do not send PrincipalOrgID

kagahd committed:
handle assumed role

kagahd committed:
adapt check to follow best practices by avoiding "NotPrincipal" in a Deny statement of a resource-based policy

kagahd committed:
make black happy

kagahd committed:
make flake8 happy

kagahd committed:
add Unit tests

jfagoagas commented:
Hello @kagahd I apologise because this PR got lost in our inbox. The team is going to review it as soon as possible.

Thanks!

kagahd commented on prowler/providers/aws/services/secretsmanager/secretsmanager_has_restrictive_resource_policy/secretsmanager_has_restrictive_resource_policy.py:
The check is about to change because of new use cases we were made aware of. Once integrated, I'll push it here.

kagahd committed:
Enhance secretsmanager_has_restrictive_resource_policy check

- Add cross-account access detection in Allow statements
- Implement validation for ArnNotLike condition patterns with minimum 12-char prefix
- Refactor Deny statement validation with stricter condition checks (case 1: both PrincipalArn and PrincipalServiceName with StringNotEqualsIfExists + Null; case 2: only PrincipalArn with StringNotEquals)
- Add detailed error messages with up to 3 principals/services displayed
- Integrate is_condition_block_restrictive from IAM policy utilities for consistent validation
- Add improved handling for wildcard principals and service-based access
- Include comprehensive validation for Allow statements with Service Principals

kagahd committed:
chore(aws): adapt metadata to new standard per metadata guidelines

kagahd committed:
chore(aws): add changelog entry for secretsmanager_has_restrictive_resource_policy

kagahd commented on prowler/providers/aws/services/secretsmanager/secretsmanager_has_restrictive_resource_policy/secretsmanager_has_restrictive_resource_policy.py:
```diff
@@ -0,0 +1,531 @@
+                cross_account_principals = []
+
+                # Process all statements
+                for statement in statements:
```
You're absolutely right — this is a bug. The current implementation processes Allow and Deny statements in a single pass, and the `break` on line 217/229 exits the loop as soon as a valid Deny is found. Any Allow statement with a cross-account principal that appears *after* that Deny is never evaluated, leading to a false PASS.

I'll refactor this to decouple the two concerns: first, collect all cross-account Allow principals across *all* statements, then separately validate the Deny statements. This will make the result completely order-independent, as it should be.

kagahd commented on prowler/providers/aws/services/secretsmanager/secretsmanager_has_restrictive_resource_policy/secretsmanager_has_restrictive_resource_policy.py:
```diff
@@ -0,0 +1,531 @@
+                                has_wildcard_action = any(
+                                    "*" in action for action in actions
+                                )
+                                has_correct_condition = condition == {
```
Good catch. The current check (`condition == {"StringEquals": {"aws:SourceAccount": ...}}`) rejects policies that add *additional* restrictive conditions like `ArnLike` on `aws:SourceArn`. That's overly strict — a policy that is *more* restrictive than the baseline should still pass.

I'll switch this to a subset/superset check: verify that the required `StringEquals` / `aws:SourceAccount` condition is *present*, without rejecting additional conditions that only tighten the policy further.

kagahd commented on prowler/providers/aws/services/secretsmanager/secretsmanager_has_restrictive_resource_policy/secretsmanager_has_restrictive_resource_policy.py:
```diff
@@ -0,0 +1,531 @@
+                failed_principals = []
+                failed_services = []
+
+                # Validate that NotAction does not contain wildcards for specified principals
```
This is an intentional design decision: the per-principal Deny/NotAction statements are considered an *optional hardening layer* on top of the required baseline (which is the global Deny with `StringNotEquals` / `aws:PrincipalArn`). The rationale is that the global Deny already restricts access to only the listed principals — the per-principal NotAction blocks are an additional defense-in-depth measure that limits *what* each allowed principal can do, but they're not strictly required for the policy to be considered "restrictive."

I fully agree this needs to be documented explicitly though. I'll add a clear comment in the code and a note in the metadata explaining this design choice.

kagahd commented on prowler/providers/aws/services/secretsmanager/secretsmanager_has_restrictive_resource_policy/secretsmanager_has_restrictive_resource_policy.py:
```diff
@@ -0,0 +1,531 @@
+        # Regular expression to match IAM roles or users without wildcard * in their name
+        arn_pattern = r"arn:aws:iam::\d{12}:(role|user)/([^*]+)$"
+        # Regular expression to match AWS service names
+        service_pattern = r"^[a-z0-9-]+\.amazonaws\.com$"
```
Fair point. The current regex only matches simple, non-regionalized principals like `appflow.amazonaws.com`. For now, this is intentional — the services we've encountered in practice with Secrets Manager resource policies all use simple principals.

However, I'll document this constraint explicitly and adjust the regex to also handle regionalized and multi-label principals (e.g., `logs.eu-central-1.amazonaws.com`) to be future-proof.

kagahd commented:
Hi @HugoPBrito,

Thanks a lot for the detailed and well-researched review. It's clear you spent real time understanding the logic, and I really appreciate that, especially given the complexity of this check. I'll push fixes for points 1 and 2 as priority, update the documentation for point 3, and widen the regex for point 4.

kagahd committed:
fix(aws): address review feedback for secretsmanager_has_restrictive_resource_policy

- Decouple Allow/Deny scanning into two passes for order-independent
  cross-account detection (fixes false PASS when Allow follows Deny)
- Switch service Allow condition check from exact-equality to subset
  check so additional restrictive conditions are accepted
- Widen service principal regex to support regionalized principals
  (e.g. logs.eu-central-1.amazonaws.com)
- Document NotAction optionality as intentional design decision
- Update metadata description, recommendation, and notes accordingly
- Add regression tests for all three code changes

kagahd commented:
Hi @HugoPBrito,

All four points from your review have been addressed in commit f26cba1:

1. **Loop ordering bug (line 79)** - Decoupled Allow/Deny scanning into two separate passes. Cross-account Allow principals are now collected independently before Deny validation, making the result fully order-independent. Added a regression test that places a cross-account Allow after the Deny blocks.

2. **Condition exact-equality (line 354)** - Switched from exact-equality to a subset check. The condition now verifies that `StringEquals` with `aws:SourceAccount` is present, without rejecting additional restrictive conditions like `ArnLike`. Added a test with extra conditions that confirms PASS.

3. **NotAction optionality (line 319)** - Documented the design decision in a code comment and updated the metadata (Description, Recommendation, Notes) to clearly state that per-principal Deny/NotAction statements are an optional defense-in-depth layer.

4. **Service principal regex (line 18)** - Widened the regex to accept regionalized and multi-label principals (e.g. `logs.eu-central-1.amazonaws.com`). Added a test to verify. Updated the metadata to document support for both formats.

kagahd committed:
style(aws): run black/flake8 and fix metadata description length

- Apply black formatting to .py files
- Shorten metadata description to 342 chars (limit: 400)

kagahd committed:
Clarify validation comments for resource policy checks

Updated comments to clarify validation requirements for service-principal Allow statements in resource policies.

kagahd commented:
Hi @HugoPBrito,

Just a quick ping - all four points from your review have been addressed in the latest commits. Would you have a chance to take another look when you get a moment? Happy to make any further adjustments if needed.

Thanks!

HugoPBrito commented:
> Hi @HugoPBrito,
> 
> Just a quick ping - all four points from your review have been addressed in the latest commits. Would you have a chance to take another look when you get a moment? Happy to make any further adjustments if needed.
> 
> Thanks!

Thanks @kagahd ! I'll be taking a look today.

HugoPBrito commented on prowler/providers/aws/services/secretsmanager/secretsmanager_has_restrictive_resource_policy/secretsmanager_has_restrictive_resource_policy.py:
```diff
@@ -0,0 +1,532 @@
+            )
+
+            if secret.policy:
+                statements = secret.policy.get("Statement", [])
```
`Statement` in IAM policies can be either a single object or a list. Here we iterate over `secret.policy.get("Statement", [])` assuming it is always a list, so a valid policy with a single statement would make us iterate over dict keys and then fail on `statement.get(...)`. Please normalize `Statement` to a list first, like we already do in other policy helpers.

kagahd replied:
Thanks for catching this! You are right, I have added the defensive normalization now, consistent with the pattern used in other checks (e.g. `organizations_scp_check_deny_regions`). Please verify in 1e86400c3.

HugoPBrito commented on prowler/providers/aws/services/secretsmanager/secretsmanager_has_restrictive_resource_policy/secretsmanager_has_restrictive_resource_policy.py:
```diff
@@ -0,0 +1,532 @@
+                        for service in principals:
+                            if service in not_denied_services:
+                                condition = statement.get("Condition", {})
+                                actions = self.extract_field(
```
This validation only checks that `Action` does not contain wildcards, but it does not require `Action` to exist at all. That means a service-principal `Allow` statement using `NotAction` plus `StringEquals/aws:SourceAccount` could still pass here, even though it would be broader than the contract described in the PR. I think we should explicitly require `Action` to be present and reject `NotAction` for these service `Allow` statements.

kagahd replied:
Good catch. I have extracted the service-principal Allow validation into a dedicated method `_validate_service_allow_statement` and it now explicitly rejects both `NotAction` and `NotResource` (since `NotResource` has the same scope-broadening effect). If neither is present, it also requires `Action` to be explicitly set. Please verify in 1e86400c3.

HugoPBrito commented on prowler/providers/aws/services/secretsmanager/secretsmanager_has_restrictive_resource_policy/secretsmanager_has_restrictive_resource_policy.py:
```diff
@@ -0,0 +1,532 @@
+                                # StringEquals with aws:SourceAccount matching the
+                                # audited account. Additional restrictive conditions
+                                # (e.g. ArnLike on aws:SourceArn) are acceptable.
+                                has_correct_condition = (
```
IAM condition keys are case-insensitive, but this check is doing an exact lookup for `aws:SourceAccount`. A policy using `AWS:SourceAccount` or `aws:sourceaccount` would be semantically valid and still fail here. We should normalize the condition keys before reading them, or reuse the shared IAM policy helper logic that already handles this correctly.

kagahd replied:
Valid point. I have added a `_normalize_condition_keys` method that lowercases all condition keys upfront, following the same approach as `is_condition_block_restrictive()` in `prowler/providers/aws/services/iam/lib/policy.py`. All internal comparisons now use lowercase keys. Please verify in 1e86400c3.

HugoPBrito commented on prowler/providers/aws/services/secretsmanager/secretsmanager_has_restrictive_resource_policy/secretsmanager_has_restrictive_resource_policy.py:
```diff
@@ -0,0 +1,532 @@
+
+                # Allow-Statement for not-denied services must not have any wildcards in "Action"
+                # and "SourceAccount" must be the audited account
+                for statement in statements:
```
Non-blocking, but I think this service-principal `Allow` validation block should be extracted into a shared IAM-policy helper, or at least a dedicated private method. Right now this check is implementing its own condition-parsing semantics instead of leaning on the common policy utilities, and that is already causing drift in behavior. Pulling it out would make the check easier to reason about and reduce the chance of policy-validation bugs creeping back in.

kagahd replied:
Agreed. I have extracted the service-principal Allow validation into `_validate_service_allow_statement` as a private method. Moving it into a shared IAM policy helper would be a reasonable next step, but for now this at least isolates the logic and makes it easier to reason about. Please verify in 1e86400c3.

HugoPBrito reviewed (commented):
Sorry for the amount of requested changes throughout this PR review. I know this has been a long and detailed iteration, and I really appreciate the effort you have put into it, @kagahd. Thanks for staying with it and addressing the previous rounds so carefully. I have a few final review comments below.

kagahd committed:
fix(aws): harden policy validation for secretsmanager_has_restrictive_resource_policy

- Normalize Statement to list (IAM spec allows single dict)
- Normalize condition keys to lowercase (case-insensitive per AWS spec)
- Reject NotAction and NotResource in service-principal Allow statements
- Extract service Allow validation into _validate_service_allow_statement

kagahd commented:
@HugoPBrito, there is no need to apologize! Having a thorough reviewer on a check this complex is genuinely valuable. Your comments have consistently improved the robustness of the implementation (the case-insensitive condition keys and the NotAction gap were real bugs that would have slipped through). Appreciate the detailed feedback, it makes the end result better.

kagahd committed:
fix(aws): flatten mixed-principal dicts and accept SourceAccount as list

extract_field() now returns both AWS and Service values from dict
Principals instead of only the first branch found. This fixes a bug
where mixed-principal Allow statements like {"AWS": "...", "Service":
"appflow.amazonaws.com"} silently skipped service-principal validation.

Also support aws:SourceAccount as a single-value list in StringEquals
conditions, matching AWS IAM behavior.

kagahd committed:
fix(aws): make secretsmanager resource policy check partition-aware

All regex patterns were hardcoded to the commercial partition (arn:aws:,
.amazonaws.com), causing false FAILs in aws-cn and aws-us-gov. Now uses
secretsmanager_client.audited_partition to build patterns dynamically:

- ARN patterns (arn_pattern, iam_arn_with_account_pattern,
  iam_root_arn_pattern, arn_wildcard_pattern) use the actual partition
- Service principal pattern uses .amazonaws.com.cn for aws-cn,
  .amazonaws.com for all other partitions
- Assumed-role ARN transformation (sts -> iam) respects the partition

HugoPBrito commented:
> Thanks for the green light @HugoPBrito! Happy to provide evidences. To make sure I share the right thing: would sanitized output from a test run against a sandbox account (one secret with a compliant policy, one without) work for you? I'd rather not post production findings for obvious reasons. Alternatively, the 80 unit tests cover all validation paths if that's sufficient.
> 
> Regarding the `threat-detection` category: I'm not sure it fits here since this is a static configuration compliance check, not an active threat detection mechanism. Adding it to that category could set the wrong expectation about what the check does. If execution time is a concern, I'm happy to look into performance optimizations or discuss whether a dedicated scan profile would be a better fit. What are your thoughts?

Sure! That'll be fine.

Regarding performance, feel free to give it a try. If not, I can also take a look into it. It concerns me because it could take a really long times in big accounts and I'm worried about Prowler Cloud customers.

HugoPBrito reviewed (changes requested):
Please change that and it'll be at v5.25! If not done by the needed time I'll do it myself so it's included in the release.

kagahd committed:
chore: move changelog entry to v5.25.0 section

kagahd committed:
merge upstream/master and resolve CHANGELOG conflict

kagahd committed:
fix: remove remaining conflict marker in CHANGELOG.md
